### DISCUSSION TITLE
RFC: new formState `isValidating`

### DISCUSSION THREAD

**Discussion by bluebill1049** [react-hook-form/react-hook-form#3669] [RFC]
## Context

React Hook Form validates input async by default, this is great which gives us a lot of flexibility in terms of validation and potentially easily integrates with any server-side validation. However, there is a price to pay that as well which means every action or user interaction will result in async too, right now we are lack of `formState` to support such state update, which appeared on the following feature request: 

https://github.com/react-hook-form/react-hook-form/discussions/2841
https://github.com/react-hook-form/react-hook-form/issues/3627

#### Pros
- users can subscribe to the form state while input is getting validate async

#### Cons
- increase in package bundle size, every feature there is a price to pay and  `formState` update is always going to be a global/core update and can't be tree shaken, 

## Proposal

new boolean state `isValidating`

**Option A**

```tsx
const { formState: { isValidating } } = useForm();
const { meta: { isValidating } } =  useController();
```

**Option B**

```tsx
const { status: 'validating' } = useForm();
const { status: 'validating' } =  useController();
```

## Implementation

This change will most likely to impact on `useForm` only (with my quick spike). FormState will be guarded with proxy as well, so if you are not reading this state, it will not trigger extra render for your form.

## Poll result (from the community):
https://twitter.com/HookForm/status/1339155959076122624

<img width="646" alt="Screen Shot 2020-12-18 at 9 45 46 pm" src="https://user-images.githubusercontent.com/10513364/102606126-64546480-417a-11eb-883f-b6c21e3bb7e6.png">

There are demands with such form state, but not as popular as other states.

**Comment by charleslevesque** [comment]
Like [@ArnBZN said on Twitter](https://twitter.com/HookForm/status/1339155959076122624), with an impact on the size with less than 10% and the complexity is stable, I think this will be a nice addition to the library because it would open more ways to display information to the user specifically with long-running server-side validations. It also enables a more fine-grained control over the state of the form like canceling a submission while it’s validating before mutating the data.

Option A to me seems the more consistent approach with other properties available and looks as if it will also be a lesser effort to integrate into the documentation.

**Reply by bluebill1049** [reply]
Thanks, @charleslevesque for your feedback. I do agree some use cases actually do require `isValdiating` state.

**Comment by emilio-martinez** [comment]
As far as approach goes, since the validation happens at a field level, would it not be appropriate to show this with that field granularity? For example, if you have a username field that needs to check of uniqueness you might want to indicate that next to the field

**Reply by bluebill1049** [reply]
It will have to be a global `formState`.

**Comment by bluebill1049** [comment]
Went live with https://github.com/react-hook-form/react-hook-form/releases/tag/v6.14.0
